src/main/java/org/reaktivity/nukleus/kafka/internal/cache/KafkaCacheCursorFactory.java

```diff
@@ -491,40 +495,64 @@ public long reset(
 
                 if (segment != null)
                 {
-                    final KafkaCacheIndexFile indexFile = segment.indexFile();
-                    assert indexFile != null;
+                    final KafkaCacheIndexFile hashFile = segment.hashFile();
+                    assert hashFile != null;
 
-                    this.indexFile = indexFile;
+                    this.hashFile = hashFile;
 
-                    final int offsetDelta = (int)(offset - segment.baseOffset());
-                    cursor = indexFile.first(offsetDelta);
+                    if (position == POSITION_UNSET)
+                    {
+                        final KafkaCacheIndexFile indexFile = segment.indexFile();
+                        assert indexFile != null;
+                        final int offsetDelta = (int)(offset - segment.baseOffset());
+                        position = cursorValue(indexFile.first(offsetDelta));
+                    }
+
+                    cursor = hashFile.first(hash);
+                    if (cursorValue(cursor) != cursorValue(RETRY_SEGMENT))
+                    {
+                        final int cursorIndex = cursorIndex(cursor);
+                        final long cursorFirstHashWithPosition = cursor(cursorIndex, position);
+                        cursor = hashFile.ceiling(hash, cursorFirstHashWithPosition);
+                    }
                 }
                 else
                 {
-                    this.indexFile = null;
+                    this.hashFile = null;
                 }
 
                 return cursor;
             }
 
             @Override
-            public long next(
+            public final long next(
                 long cursor)
             {
-                return indexFile != null ? indexFile.resolve(cursor) : NEXT_SEGMENT;
+                return hashFile != null ? hashFile.resolve(cursor) : NEXT_SEGMENT;
```

Comment:
This will advance to the next position with matching hash, meaning next position with candidate header match.
There is potential for this to skip over offsets that do not have a header match, which should be returned by this header-not condition.
Similar feedback applies to `reset` above which starts at the first position with candidate header match.